src/libraries/System.Diagnostics.Tracing/tests/BasicEventSourceTest/ActivityTracking.cs

```diff
@@ -0,0 +1,154 @@
+            // that the thread will be sheduled to run one of our other tasks
+            // and we can detect the non-zero id at the start of the task
+            List<Task> tasks = new List<Task>();
+            for (int i = 0; i < 100; i++)
+            {
+                tasks.Add(Task.Run(() => YieldTwoActivitiesDeep(es)));
+            }
+
+            await Task.WhenAll(tasks.ToArray());
+        }
+
+        private async Task YieldTwoActivitiesDeep(ActivityEventSource es)
+        {
+            Assert.Equal(Guid.Empty, EventSource.CurrentThreadActivityId);
+            es.ExampleStart();
+            es.Example2Start();
+            await Task.Yield();
+            es.Example2Stop();
+            es.ExampleStop();
+            Assert.Equal(Guid.Empty, EventSource.CurrentThreadActivityId);
+        }
+
+        // I don't know if this behavior is by-design or accidental. For now
+        // I am attempting to preserve it to lower back compat risk, but in
+        // the future we might decide it wasn't even desirable to begin with.
+        // Compare with SetCurrentActivityIdAfterEventDoesNotFlowAsync below.
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotBrowser))]
+        public async Task SetCurrentActivityIdBeforeEventFlowsAsync()
+        {
+            using ActivityEventListener l = new ActivityEventListener();
+            using ActivityEventSource es = new ActivityEventSource();
+            try
+            {
+                Guid g = Guid.NewGuid();
+                EventSource.SetCurrentThreadActivityId(g);
+                Assert.Equal(g, EventSource.CurrentThreadActivityId);
+                es.ExampleStart();
+                await Task.Yield();
+                es.ExampleStop();
+                Assert.Equal(g, EventSource.CurrentThreadActivityId);
+            }
+            finally
+            {
+                EventSource.SetCurrentThreadActivityId(Guid.Empty);
+            }
+        }
+
+        // I don't know if this behavior is by-design or accidental. For now
+        // I am attempting to preserve it to lower back compat risk, but in
+        // the future we might decide it wasn't even desirable to begin with.
+        // Compare with SetCurrentActivityIdBeforeEventFlowsAsync above.
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotBrowser))]
+        public async Task SetCurrentActivityIdAfterEventDoesNotFlowAsync()
+        {
+            using ActivityEventListener l = new ActivityEventListener();
+            using ActivityEventSource es = new ActivityEventSource();
+            try
+            {
+                es.ExampleStart();
+                Guid g = Guid.NewGuid();
+                EventSource.SetCurrentThreadActivityId(g);
+                Assert.Equal(g, EventSource.CurrentThreadActivityId);
+                await Task.Yield();
+                Assert.NotEqual(g, EventSource.CurrentThreadActivityId);
```

Comment:
Say you get rescheduled on the same thread, do we still expect this to not be g?

Reply:
Correct, I still don't expect it to be g as long as the scheduler switches out the ExecutionContext and then switches it back in (which I think yield always does but I am not certain). EventSource.SetCurrentThreadActivityId() saves g into a thread-local storage location, but it doesn't modify the async-local. When the scheduler switches the ExecutionContext back in whatever value is stored in the async-local will be written over the thread-local storage location, erasing g.

Reply:
> which I think yield always does but I am not certain

Same understanding.

>  EventSource.SetCurrentThreadActivityId saves into a thread-local storage location, but it doesn't modify the async-local.

Ah, this will bite me for life.